spark3/src/main/java/org/apache/iceberg/spark/source/SparkMicroBatchStream.java

```diff
@@ -0,0 +1,310 @@
+    StreamingOffset batchStartOffset = StreamingOffset.START_OFFSET.equals(startOffset) ?
+        new StreamingOffset(SnapshotUtil.oldestSnapshot(table).snapshotId(), 0, false) :
+        startOffset;
+
+    do {
+      final StreamingOffset currentOffset =
+          latestMicroBatch != null && latestMicroBatch.lastIndexOfSnapshot() ?
+          getNextAvailableSnapshot(latestMicroBatch.snapshotId()) :
+              batchStartOffset;
+
+      latestMicroBatch = MicroBatches.from(table.snapshot(currentOffset.snapshotId()), table.io())
+          .caseSensitive(caseSensitive)
+          .specsById(table.specs())
+          .generate(currentOffset.position(), Long.MAX_VALUE, currentOffset.shouldScanAllFiles());
+
+      fileScanTasks.addAll(latestMicroBatch.tasks());
+    } while (latestMicroBatch.snapshotId() != endOffset.snapshotId());
+
+    return fileScanTasks;
+  }
+
+  private StreamingOffset getNextAvailableSnapshot(long snapshotId) {
+    Snapshot previousSnapshot = table.snapshot(snapshotId);
+    Snapshot pointer = table.currentSnapshot();
+    while (pointer != null && previousSnapshot.snapshotId() != pointer.parentId()) {
+      Preconditions.checkState(pointer.operation().equals(DataOperations.APPEND),
+          "Encountered Snapshot DataOperation other than APPEND.");
+
+      pointer = table.snapshot(pointer.parentId());
+    }
+
+    Preconditions.checkState(pointer != null,
+        "Cannot read data from snapshot which has already expired: %s", snapshotId);
+
+    return new StreamingOffset(pointer.snapshotId(), 0L, false);
+  }
+
+  interface InitialOffsetStore {
+    static InitialOffsetStore getInstance(Table table, String checkpointLocation) {
+      return new InitialOffsetStoreImpl(table, checkpointLocation);
+    }
+
+    StreamingOffset addInitialOffset();
+
+    boolean isOffsetStoreInitialized();
+
+    StreamingOffset getInitialOffset();
+  }
+
+  private static class InitialOffsetStoreImpl implements InitialOffsetStore {
+    private final Table table;
+    private final FileIO io;
+    private final String initialOffsetLocation;
+
+    InitialOffsetStoreImpl(Table table, String checkpointLocation) {
+      this.table = table;
+      this.io = table.io();
+      this.initialOffsetLocation = new Path(checkpointLocation, "offsets/0").toString();
+    }
+
+    @Override
+    public StreamingOffset addInitialOffset() {
+      table.refresh();
+      StreamingOffset offset = table.currentSnapshot() == null ?
+          StreamingOffset.START_OFFSET :
+          new StreamingOffset(SnapshotUtil.oldestSnapshot(table).snapshotId(), 0, true);
+
+      OutputFile file = io.newOutputFile(initialOffsetLocation);
+      serialize(offset, file);
+
+      return offset;
+    }
+
+    @Override
+    public boolean isOffsetStoreInitialized() {
+      InputFile file = io.newInputFile(initialOffsetLocation);
+      return file.exists();
+    }
+
+    @Override
+    public StreamingOffset getInitialOffset() {
```

Comment:
I think I mentioned this elsewhere, but `get` in a method name is a red flag to me that there is either a better verb or that the method name could be shorter by simply omitting `get`. There are very few cases where `get` in a method name is a good idea. The two I can think of are:
1. Fetching a singleton instance, like `IntegerType.get()`
2. The method is a getter _and_ the class must conform to Java bean conventions

Otherwise, I think `get` should be replaced or omitted.